README.md
hapi-dashboard
==============

Hapi plugin for displaying metrics on a dashboard


## Installation

`npm install --save hapi-dashboard`

## Usage

```js
server.pack.register([
  { plugin: require('hapi-dashboard'), options: {
    //defaults shown below
    endpoint: '/dashboard'
    dashboards: {},
    ttl: 60*1000, //time to cache responses. 60 seconds
    concurrent: 20, //max metrics to process at once
    password: 'something', //password used to access dashboards
	favicon: 'favicon.png' // Also used as home screen icon for ios
  }}
], function(err) {
});
```

## Dashboards Object

Dashboards object should look something like this:

```js
{
  'dashboard-1': {
    name: 'Dashboard 1',
    metrics: {
      loggedIn: {
        value: function(server, done) {
          //perform some async call
          server.plugins.db.metrics.loggedInCount(function(err, result) {
            done(null, {
            name: 'Number of times a user logged in',
            value: result.total
          });
          }
        },
        ttl: 60
      }
    }
  }
}
```


CHANGELOG.md

1.3.0 / 2017-02-02
==================

  * allow for more complex caching
  * chore(package): update dependencies

1.2.3 / 2017-01-31
==================

  * set method loader to not be verbose

1.2.2 / 2017-01-31
==================

  * set default expiresIn

1.2.1 / 2017-01-31
==================

  * removed hapi-password from example
  * fixed caching
  * updated deps

1.2.0 / 2017-01-31
==================

  * continue if error

1.1.0 / 2017-01-31
==================

  * log in example
  * output duration in log

1.0.0 / 2016-12-19
==================

  * es6 updates

0.7.2 / 2016-05-12
==================

  * this is why you should test things

0.7.1 / 2016-05-12
==================

  * added generateTimeout

0.7.0 / 2015-07-07
==================

  * log when fetching data
  * use exports name as key
  * if undefined, set to '' in csv
  * added charset utf8 on views

0.6.3 / 2015-01-23 
==================

  * fix for csv output

0.6.2 / 2015-01-09
==================

  * Fixed backwards compatibility

0.6.1 / 2015-01-07
==================

  * Cleaned up normalize code and fixed csv export.

0.6.0 / 2015-01-05
==================

  * Normalize data to array.
  * Allows data to be ordered in an array

0.5.0 / 2014-11-21 
==================

  * Added request object to `this` in value function.

0.4.1 / 2014-11-04 
==================

  * fixed auth strategy defaults to allow null to work

0.4.0 / 2014-11-04 
==================

  * Added ability to group metrics
  * Defaults expiresIn to undefined.
  * Allows root endpoint to be set.
  * Setup auth in server rather than plugin.
  * Removed old authentication - added support for auth strategies.

0.3.0 / 2014-10-14 
==================

  * default to no password.  fixes #22
  * removed plugin.ext and replaced with passing in options to view

0.2.0 / 2014-10-14 
==================

  * Allow querystring auth
  * Added csv and json exporters
  * Added scripts to pages
  * Added favicon support
  * Added viewport tags and fixed all dashboards link
  * Added css support to login form
  * Basic password protection

0.1.0 / 2014-10-07 
==================

  * Supports html, custom columns, array of metric data

0.0.2 / 2014-10-06 
==================

  * added support to pass in cssFiles to the dashboard views
  * fixed paths to views and partials

0.0.1 / 2014-10-06 
==================

  * Initial commit


example/server.js
/* eslint no-console: 0 */

'use strict';

const Hapi = require('hapi');
const server = new Hapi.Server({
  debug: {
    request: ['*'],
    log: ['hapi-dashboard']
  }
});

server.connection({
  port: 8000,
});

const dashboardOptions = {
  favicon: 'http://placekitten.com/g/152/152',
  endpoint: '/',
  cache: {
    expiresIn: 1000 * 60,
    staleIn: 1000,
    staleTimeout: 10,
    generateTimeout: 5 * 1000
  },
  dashboards: {
    users: {
      name: 'Users',
      metrics: {
        adminUsers: {
          section: 'Admin',
          value(srv, done) {
            setTimeout(() => {
              done(null, {
                Name: 'Admin Users',
                Total: 12
              });
            }, 1000);
          }
        },
        normalUsers: {
          section: 'User',
          value(srv, done) {
            setTimeout(() => {
              done(null, {
                Name: 'Normal Users',
                Total: 2734
              });
            }, 30);
          }
        },
        bannedUsers: {
          section: 'User',
          value(srv, done) {
            setTimeout(() => {
              done(null, {
                Name: 'Banned Users',
                Total: 1
              });
            }, 70);
          }
        }
      }
    },
    performance: {
      name: 'Performance',
      metrics: {
        pageViews: {
          value(srv, done) {
            setTimeout(() => {
              done(null, {
                name: 'Page Views',
                Total: 87643245
              });
            }, 40);
          },
          ttl: 0
        },
        bounceRate: {
          value(srv, done) {
            setTimeout(() => {
              done(null, {
                name: 'Bounce Rate',
                Total: '99%'
              });
            }, 30);
          },
          ttl: 0
        }
      }
    },
    errors: {
      name: 'App Errors',
      metrics: {
        appErrors: {
          value(srv, done) {
            done(null, [
              {
                name: 'Internal Errors',
                value: 1234
              },
              {
                name: 'Email Errors',
                value: 1234
              }
            ]);
          }
        }
      }
    },
    randomdata: {
      name: 'Random Data',
      metrics: {
        something: {
          value(srv, done) {
            done(null, [
              {
                ' ': 'cat',
                before: '<img src="http://placekitten.com/g/200/300">',
                after: '<img src="http://placekitten.com/g/50/50">'
              },
              {
                ' ': 'meat',
                before: '<img src="http://baconmockup.com/300/200">',
                after: '<img src="http://baconmockup.com/50/50">'
              }
            ]);
          }
        }
      }
    },
    requeststuff: {
      name: 'Request Stuffs',
      metrics: {
        request: {
          value(srv, done) {
            done(null, [this.headers]);
          }
        }
      }
    },
    daily: {
      name: 'Daily',
      metrics: {
        daily: {
          value(srv, done) {
            done(null, [
              {
                name: 'Search',
                data: [
                  {
                    name: '01/05/15',
                    value: 3
                  },
                  {
                    name: '01/04/15',
                    value: 1
                  },
                  {
                    name: '01/03/15',
                    value: 18
                  },
                  {
                    name: '01/02/15',
                    value: 5
                  },
                  {
                    name: '01/01/15',
                    value: 1
                  },
                  {
                    name: '12/31/14',
                    value: 764
                  }
                ]
              }
            ]);
          }
        }
      }
    }
  },
};

server.register([
  require('vision'),
  {
    register: require('../'),
    options: dashboardOptions
  }
], err => {
  if (err) {
    console.error(err);
  }

  server.start(() => {
    console.log('Server started at:', server.info.uri);
  });
});
